metricflow/cli/main.py

```diff
@@ -721,5 +733,181 @@ def validate_configs(cfg: CLIContext, dw_timeout: Optional[int] = None, skip_dw:
+    solver = staticmethod(_get_spin_context_manager_factory("🧠 Solving column types"))  # type: ignore
+    renderers = staticmethod(_get_spin_context_manager_factory("📝 Writing output"))  # type: ignore
+
+
+@cli.command()
+@click.option(
+    "--tables",
+    cls=click_custom.MutuallyExclusiveOption,
+    mutually_exclusive=["schema"],
+    type=click_custom.ListParamType(
+        value_converter=lambda table_str: SqlTable.from_string(table_str),
+        min_length=1,
+    ),
+    required=False,
+    help="Comma-separated list of table names to be queried for inference.",
+)
+@click.option(
+    "--schema",
+    cls=click_custom.MutuallyExclusiveOption,
+    mutually_exclusive=["tables"],
+    type=str,
+    required=False,
+    help="Name of a schema to be queried for inference. Will list all tables in this schema.",
+)
+@click.option(
+    "--max-sample-size",
+    type=click.IntRange(min=1),
+    required=True,
+    default=10000,
+    help="The maximum number of rows to sample from the warehouse, for each table.",
+)
+@click.option(
+    "--solver-threshold",
+    type=click.FloatRange(min=0.5, max=1),
+    required=False,
+    default=0.6,
+    help="The inference solver weight threshold.",
+)
+@click.option(
+    "--solver-weights",
+    type=click_custom.ListParamType(
+        value_converter=lambda weight_str: int(weight_str),
+        min_length=4,
+        max_length=4,
+    ),
+    required=False,
+    default="1,2,4,8",
+    help="Comma-separated list of 4 integer weights to be assigned for each confidence value. "
+    "Interpreted weights will be assigned to confidences LOW,MEDIUM,HIGH,VERY_HIGH",
+)
+@click.option(
+    "--output-dir",
+    type=click.Path(
+        dir_okay=True,
+        file_okay=False,
+        writable=True,
+        readable=True,
+    ),
+    required=True,
+    help="Output directory for the inferred config files.",
+)
+@click.option(
+    "--overwrite",
+    is_flag=True,
+    required=True,
+    default=False,
+    help="If specified, allows existing configuration files to be overwritten by inference.",
+)
+@beta_feature_warning(feature_name="Data Source Inference")
+@pass_config
+@exception_handler
+@log_call(module_name=__name__, telemetry_reporter=_telemetry_reporter)
+def infer(
+    cfg: CLIContext,
+    tables: Optional[List[SqlTable]],
+    schema: Optional[str],
+    max_sample_size: int,
+    solver_threshold: float,
+    solver_weights: List[int],
+    output_dir: str,
+    overwrite: bool,
+) -> None:
+    """Infer data source configurations from warehouse information."""
+
+    if cfg.sql_client.sql_engine_attributes.sql_engine_type is not SupportedSqlEngine.SNOWFLAKE:
+        click.echo(
+            "Data Source Inference is currently only supported for Snowflake. "
+            "We will add support for all the other warehouses before it becomes a "
+            "stable feature. Stay tuned!"
+        )
+        return
+
+    if tables is None and schema is None:
```

Comment:
If you want this to have a sorta XOR operation there's a helper `assert_exactly_one_arg_set` in `metricflow/object_utils.py`. Could possibly be useful here

Reply:
that is good to know :)